harmony/taskhelp/usertaskmgt/usertaskmgt.go

```diff
@@ -0,0 +1,152 @@
+*/
+package usertaskmgt
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/samber/lo"
+	"golang.org/x/xerrors"
+)
+
+var log = logging.Logger("userTaskMgt")
+
+func WrapTasks(tasks []harmonytask.TaskInterface, UserScheduleUrl []string, db *harmonydb.DB, hostAndPort string) {
+	urlMap := lo.SliceToMap(UserScheduleUrl, func(s string) (string, *url.URL) {
+		spl := strings.SplitN(s, ",", 2)
+		if len(spl) != 2 {
+			log.Errorf("Invalid UserScheduleUrl: %s. Expected: taskName,url", s)
+			return "", &url.URL{}
+		}
+		u, err := url.Parse(spl[1])
+		if err != nil {
+			log.Errorf("Invalid UserScheduleUrl: %s. Expected: taskName,url", s)
+			return "", &url.URL{}
+		}
+		return spl[0], u
+	})
+	for i, task := range tasks {
+		if url, ok := urlMap[task.TypeDetails().Name]; ok {
+			tasks[i] = &UrlTask{
+				TaskInterface:   task,
+				UserScheduleUrl: url,
+				name:            task.TypeDetails().Name,
+				db:              db,
+				hostAndPort:     hostAndPort,
+			}
+		}
+	}
+}
+
+type UrlTask struct {
+	harmonytask.TaskInterface
+	db              *harmonydb.DB
+	UserScheduleUrl *url.URL
+	name            string
+	hostAndPort     string
+}
+
+// CanAccept should accept all IF no harmony_task_user row exists, ELSE
+// if us, try CanAccept() until expiration hits.
+func (t *UrlTask) CanAccept(tids []harmonytask.TaskID, te *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	id := tids[0]
+	var owner string
+	var expiration int64
+	err := t.db.QueryRow(context.Background(), `SELECT COALESCE(owner,''), COALESCE(expiration, 0) from harmony_task_user WHERE task_id=$1`, id).Scan(&owner, &expiration)
+	if err != nil {
+		return nil, xerrors.Errorf("could not get owner: %w", err)
+	}
+	if owner != "" {
+		if owner == t.hostAndPort {
+			return t.TaskInterface.CanAccept(tids, te)
+		}
+		if expiration < time.Now().Unix() {
+			_, err = t.db.Exec(context.Background(), `DELETE FROM harmony_task_user WHERE task_id=$1`, id)
+			if err != nil {
+				return nil, xerrors.Errorf("could not delete from harmony_task_user: %w", err)
+			}
+		}
+	}
+	return &id, nil
+}
+
+var client = &http.Client{Timeout: time.Second * 10}
+
+func (t *UrlTask) Do(id harmonytask.TaskID, stillMe func() bool) (bool, error) {
+	var owner string
+	err := t.db.QueryRow(context.Background(), `SELECT COALESCE(owner,'') FROM harmony_task_user WHERE task_id=$1`, id).Scan(&owner)
+	if err != nil {
+		return false, xerrors.Errorf("could not get owner: %w", err)
+	}
+	if owner == t.hostAndPort {
+		return t.TaskInterface.Do(id, stillMe)
+	}
+	var workerList []string
+	err = t.db.Select(context.Background(), &workerList, `SELECT host_and_port 
+	FROM harmony_machines m JOIN harmony_machine_details d ON d.machine_id=m.id 
+	WHERE tasks LIKE $1`, "%,"+t.name+",%")
```

Comment:
Only thing I worry about here is that the worker list technically includes workers which won't be able to take the work, but really that's probably fine.

Reply:
We have an expiration of the "appointment" to handle the extremes of that situation.